aws/internal/service/ec2/waiter/status.go

```diff
@@ -245,6 +245,32 @@ func InstanceIamInstanceProfile(conn *ec2.EC2, id string) resource.StateRefreshF
 	}
 }
 
+const (
+	ErrCodeInvalidRouteTableIDNotFound = "InvalidRouteTableID.NotFound"
+
+	RouteTableStatusReady = "ready"
+)
+
+func RouteTableStatus(conn *ec2.EC2, id string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		output, err := finder.RouteTableByID(conn, id)
+
+		if tfawserr.ErrCodeEquals(err, ErrCodeInvalidRouteTableIDNotFound) {
```

Comment:
Should be `if tfresource.NotFound(err)` as `RouteTableByID` returns a `resource.NotFoundError` if no route is found.

Reply:
AWS has some consistency problems with the ec2 package. However, I'm uncomfortable changing the API's response in the finder to `resource.NotFoundError`. At some point, the API will/should add the error const. Between the options here, changing the response to `NotFoundError` and returning the API response and using our own const, the tail on latter option is less problematic. We'd simply swap our const for their const. However, when changing the API response, we're not always sure future will code will use the finder leading to the problem of handling separate errors from the supposedly same API call. Also, it means the route table code is inconsistent using `tfresource.NotFound()` instead of `tfawserr.ErrCodeEquals()` like most other places we check `d.IsNewResource()`. Thoughts?

Reply:
1. Agree that consistency is important
2. Right now _most_ finders returns `nil`, `nil` when the resource is not found although _some_ are now returning `nil`, `resource.NotFoundError` (see #17613, #18433)
3. `RouteTableByID` is already called in a number of replaces and is returning `nil`, `resource.NotFoundError`
4. As a totally personal opinion, I have come round from thinking that `nil`, `nil` is better to preferring `nil`, `resource.NotFoundError` (better error propagation, less possibility for `nil` pointer crashes)

Reply:
I think this needs more discussion and particularly to get everyone on board one way or the other. Since each approach takes such a drastically different path (error vs. no error), a lot of code would need to be changed to conform to a new approach. It will be confusing and error prone to have some code take one path and other take another. Casual contributors will randomly grab one or the other.

Should the pattern be 1) for finders to return an error for not found, and 2) for the waiter to eat the error when appropriate (`Deleted` waiters)? Then you've got code that uses a finder handling an error while code using the waiter doesn't.

It's all doable and not necessarily right/wrong. But, in my mind, the key is for us to together figure out what we'll do so that we're on the same page and get the community on that page. To continue the metaphor, how helpful and correct a new page is should be weighed against the challenge of getting everyone on that page.

Reply:
When I first started creating the `finder` functions, I went with `nil, nil` mostly because it was simple, and I wasn't really thinking of the standard errors that the Plugin SDK provided. However, after some more thought, I figured that returning a `resource.NotFoundError` would be more semantic, allow the function to be more self-contained, and give us consistent error handling.

If you need to access the underlying AWS error, it is available in the `LastError` field of the `resource.NotFoundError`.

One additional benefit of using a `resource.NotFoundError` is avoiding checking for multiple not found cases, e.g.

https://github.com/hashicorp/terraform-provider-aws/blob/19b805575ff44160ae849a5f06b67ca9fcb26701/aws/resource_aws_backup_vault.go#L83-L96

https://github.com/hashicorp/terraform-provider-aws/blob/19b805575ff44160ae849a5f06b67ca9fcb26701/aws/resource_aws_budgets_budget.go#L311-L326

Reply:
In order to reduce creep scope in this PR, we may want to consider rewinding back to 1961ca17f4395772a99ae8fd0624f53af4410231 and moving all the subsequent commits into a new PR or multiple new PRs.